### ISSUE TITLE
live_dashboard_path/2 is undefined or private warning in prod version

### ISSUE THREAD

**Issue by sasa1977** [phoenixframework/phoenix#4080]
### Environment

* Phoenix version (mix deps): 1.5.7

### Expected behavior

The generated project shouldn't emit warnings.

### Actual behavior

When compiling in prod env, the following warning is emitted:

```
live_dashboard_path/2 is undefined or private
```

in app.html.eex

If I understand correctly this happens b/c in router a live_dashboard route is not defined in prod:

```elixir
# router.ex

if Mix.env() in [:dev, :test] do
  import Phoenix.LiveDashboard.Router

  scope "/" do
    pipe_through(:browser)
    live_dashboard("/dashboard", metrics: MySystemWeb.Telemetry)
  end
end
```

Which breaks the following code from the generated app.html.eex

```elixir
# app.html.eex

<%= if function_exported?(Routes, :live_dashboard_path, 2) do %>
  <li><%= link "LiveDashboard", to: Routes.live_dashboard_path(@conn, :home) %></li>
<% end %>
```

I'm not exactly sure what fix to recommend though. Checking at compile-time if the function is exported would be my first instinct, but I'm not sure if this would introduce unwanted compile-time deps. Another option is to copy-paste `Mix.env()` check, but that introduces a non-obvious coupling (if a dev wants to change the condition they have to do it in two different and mutually remote places, which is not obvious at all).

Either way I think this should be addressed, because the generator creates the code which might break on CI (that happened to us), and it's not clear what would be the proper way to fix this.

**Comment by striking** [comment]
I am experiencing the same issue on latest version

**Comment by josevalim** [comment]
@sasa1977 I like your function exported check. Can you please submit a PR? :)

**Comment by sasa1977** [comment]
I can give it a try. Where should this be done? I could generate a helper private function in `LayoutView`, or alternatively inline this in app.html.eex, but in this I guess I need to use `unquote` to do this at compile time. Also, we need to add `require Routes` to `LayoutView`. Is such dependency acceptable?

**Comment by josevalim** [comment]
It is fine to do it at runtime. This is meant to be example code and function_exported? checks are very lightweight (faster than invoking a function).

**Comment by josevalim** [comment]
To be clear, at runtime and in the template, as you wrote. :)

**Comment by sasa1977** [comment]
I'm not sure I understand. We're already checking if the function is exported at runtime, but that's exactly the problem, because the code which invokes `Routes.live_dashboard_path` is always present, and it generates the warning in prod. 

So one thing I mentioned was conditionally generating this link at compile time, but that would make the code uglier and introduce a possibly unwanted compile-time dependency.

**Comment by josevalim** [comment]
Oh, sorry for the confusion. I thought you were proposing to add the function_exported? but now I understand we already have it. Yeah, I am honestly not sure what is the best way forward then. Anything we do at compilation time will add a compile time dependency between the layout and the view, and I would like to avoid that. It seems the easiest option is to hardcode "/dashboard" with a comment but even then... :S

Another option is to add conditional route compilation directly to the Phoenix router, so we can do:

    if: Mix.env == :dev

That keeps the helper but doesn't generate the route.

**Comment by sasa1977** [comment]
Would it be very terrible if we use `apply` to dynamically invoke the function? This would be a simple change, and we're already checking if the function exists so we should be fine.

**Comment by josevalim** [comment]
It is just a matter of time until the compiler is smart enough to see across the apply too.

**Comment by LostKobrakai** [comment]
How about using the app env and possibly `Application.compile_env` (once generators are bumped to 1.10)? That way both router and layout view depend on config at compile time, but not on each other?

**Comment by TheFirstAvenger** [comment]
@josevalim lacking any good current alternative, is @sasa1977 suggestion of using `apply` acceptable for now to at least punt this issue until the compiler can see across `apply`? I would assume that if/when that change comes, there will be a bunch of other things like this that will all need to be considered anyway, and maybe by that time we will have come up with a systemic way of handling this. Using apply for now would help eliminate the generated ci breaking warnings at least in the short term.

**Comment by josevalim** [comment]
Perhaps a better option is to add `@compile {:no_warn_undefined, {Routes, :live_dashboard_path, 2}}` to LayoutView? 🤔

**Comment by TheFirstAvenger** [comment]
@josevalim That worked. Implemented in #4241.
